homeassistant/config_entries.py

```diff
@@ -1360,6 +1360,28 @@ async def async_finish_flow(
                 ir.async_delete_issue(self.hass, HOMEASSISTANT_DOMAIN, issue_id)
 
         if result["type"] != data_entry_flow.FlowResultType.CREATE_ENTRY:
+            if (
+                (discovery_key := flow.context.get("discovery_key"))
+                and (unique_id := flow.unique_id) is not None
+                and (
+                    entry := self.config_entries.async_entry_for_domain_unique_id(
+                        result["handler"], unique_id
+                    )
+                )
+                and entry.source == SOURCE_IGNORE
+                and (known_discovery_keys := entry.data.get("discovery_keys", []))
+                and discovery_key not in known_discovery_keys
+            ):
+                _LOGGER.debug(
+                    "Updating discovery keys for %s entry %s %s -> %s",
+                    entry.domain,
+                    unique_id,
+                    known_discovery_keys,
+                    [*known_discovery_keys, discovery_key],
+                )
+                new_discovery_keys = [*known_discovery_keys, discovery_key][-10:]
+                data = entry.data | {"discovery_keys": new_discovery_keys}
+                self.config_entries.async_update_entry(entry, data=data)
```

Comment:
When a flow is aborted, we update discovery data if there's an existing ignored entry with matching `unique_id`.

We keep more than one item of discovery data to handle cases such as:
- `cast` where a single config entry manages all Google Cast devices
- integrations which support more than one kind of discovery